### PR TITLE
Cleanup cross building

### PR DESCRIPTION
**Pull request by raboof**
Previously, because of our `akka.build.scalaVersion`, it was not possible to have a depending module build for fewer Scala versions than a dependent module (e.g. building `akka-actor` for Scala 3 but not `akka-osgi`).

The obvious alternative, using the regular `++`, has the disadvantage that you'll need to keep the patch versions of Scala versions consistent between `crossScalaVersions` and the CI scripts.

This PR introduces a custom `+~` command that fixes both these limitations: you only have to provide a version prefix (like `sbt "+~ 2.13 compile"`) and there is no more trouble when `akka-osgi` has fewer `crossScalaVersions` as `akka-actor` (because it uses `++` under the hood rather than setting `scalaVersion` directly).

### CODE REVIEW

File: project/AkkaBuild.scala [hunk 1]
```diff
@@ -286,4 +288,24 @@ object AkkaBuild {
   }
 
   def majorMinor(version: String): Option[String] = """\d+\.\d+""".r.findFirstIn(version)
+
+  // So we can `sbt "+~ 2.13 clean compile"`
```
Comment (jrudolph) [inline project/AkkaBuild.scala:292]: What's the difference here to `++`?

Comment (raboof) [reply project/AkkaBuild.scala:292]: Two differences:
* `++` also requires the patch version
* When subprojects need to be excluded, `++` needs to be specified for each command

So the `++` equivalent of the above example is `sbt "++ 2.13.7 clean" "++ 2.13.7 compile"

Comment (jrudolph) [reply project/AkkaBuild.scala:292]: Btw. the syntax for multiple commands uses `;`, so you could probably also write, `++ 2.13.7 ;clean;compile`.

Comment (raboof) [reply project/AkkaBuild.scala:292]: > That's great, can you add that as a comment? Maybe we should also create an issue with sbt to maybe get that as an upstream feature eventually.

https://discuss.lightbend.com/t/feature-request-fuzzy-switch/9288

Comment (jrudolph) [reply project/AkkaBuild.scala:292]: > Interestingly that doesn't seem to work for me (with `sbt "++ 3.1.1-RC1 clean;compile"` the non-scala3 modules get compiled anyway, leading to trouble)

Yep, indeed that aspect seems to be different.

File: project/Dependencies.scala [hunk 2]
```diff
@@ -54,25 +55,15 @@ object Dependencies {
...
-          case _                       => "0.8.0"
-        }
-      })
+    Seq(crossScalaVersions := allScalaVersions, scalaVersion := allScalaVersions.head, java8CompatVersion := {
+      CrossVersion.partialVersion(scalaVersion.value) match {
+        // java8-compat is only used in a couple of places for 2.13,
+        // it is probably possible to remove the dependency if needed.
+        case Some((3, _))            => "1.0.0"
+        case Some((2, n)) if n >= 13 => "1.0.0"
+        case _                       => "0.8.0"
```
Comment (octonato) [inline project/Dependencies.scala:64]: if only used for 2.13 code base, why we pick 0.8.0 for 2.12? This has caused so much trouble already. 

I guess there is a reason for sticking to 0.8.0. Does anyone know it? Could we make add a comment here?

Comment (octonato) [inline project/Dependencies.scala:64]: if only used for 2.13 code base, why we pick 0.8.0 for 2.12? This has caused so much trouble already. 

I guess there is a reason for sticking to 0.8.0. Does anyone know it? Could we make add a comment here?

**Review by octonato — APPROVED** [review]
LGTM

File: project/Dependencies.scala (hunk 2 above)
Comment (raboof) [reply project/Dependencies.scala:64]: I think the comment is meant to be read as "java8-compat is only used in a couple of places for 2.13, and in more places for 2.12".

**Review by pvlugter — APPROVED** [review]
Nice

File: project/Dependencies.scala (hunk 2 above)
Comment (octonato) [reply project/Dependencies.scala:64]: Ok, but why are will using scala-java8-compat  v0.8.0 for the 2.12 build. This is not clear to me and we are do this very explicitly here.

Comment (raboof) [reply project/Dependencies.scala:64]: I suspect we always used 0.8.0 for 2.12, and when we added support for 2.13 we had to update to 1.0.0 there, but didn't want to risk breaking 2.12 projects so left it at 0.8.0 for 2.12.

We could check for breaking changes between 0.8.0 and 1.0.0 and if there's nothing serious we can update this to use 1.0.0 everywhere.

**Pull request merged**